Doc: Add Ansible usage patterns (#6445)
aristanetworks/ansible-avd: docs/getting-started/ansible-usage-patterns.md

- [Ansible Role: eos_cli_config_gen](#ansible-role-eos_cli_config_gen)
- [eos_designs with eos_cli_config_gen (recommended)](#eos_designs-with-eos_cli_config_gen)
  - [AVD Design Data Model with eos_designs](#avd-design-data-model-with-eos_designs)
  - [structured_config with eos_designs](#structured_config-with-eos_designs)
  - [custom_structured_configuration Prefix](#custom_structured_configuration-prefix)

## Ansible Role: eos_cli_config_gen

The `eos_cli_config_gen` role is the EOS configuration layer of AVD. It converts structured configuration data (in YAML format) into EOS configuration syntax.

!!! warning "Not Recommended for Most Users"
    Using `eos_cli_config_gen` directly is **not the typical way** to use AVD. Most users should start with the `eos_designs` role, which provides a higher-level abstraction and automatically generates the structured configuration for you.

### Workflow

```mermaid
graph LR
    A[EOS Config Data Model] --> B{eos_cli_config_gen}
    B --> C[EOS CLI Configuration]

    click A "../../../ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/data-models.html"
```

> When using `eos_cli_config_gen` directly, you provide structured configuration using the EOS CLI Config Gen data model. The role processes this input and generates the corresponding EOS CLI configuration file.

Reviewer: `EOS CLI Config Gen` -> maybe `EOS Config` instead?

Author: ```suggestion
When using `eos_cli_config_gen` directly, you provide structured configuration using the EOS Config data model. The role processes this input and generates the corresponding EOS CLI configuration file.
```

This approach might be useful in specific cases:

- You want full control over the exact configuration
- You're working with a small number of devices
- You're migrating from manual configuration to automation
- You're using AVD for non-fabric use cases

### Example

Configuring DNS name servers using `eos_cli_config_gen` directly.

```yaml title="Input YAML"
# host_vars/leaf1.yml

> ip_name_servers:

Reviewer: This key is deprecated in 6. `ip_name_server` instead?

Author: it is indeed. need to redo all with the new key.. apologies I will do it!

Sometimes you need to override or extend the configuration generated by `eos_designs` for specific devices. The `structured_config` key allows you to inject custom structured configuration that will be merged with the auto-generated configuration.

#### Workflow

```mermaid
graph LR
    A1 --> B1
    A2 --> B4
    B4 --> C[Final Structured Configuration<br/>EOS Config Data Model]
    C --> D{eos_cli_config_gen}
    D --> E[EOS CLI Configuration]

    subgraph A[AVD Design Inputs]
        direction TB
        A1[Design Model]
        A2[structured_config in the model]
        %% This forces A1 above A2 without a visible line
        A1 ~~~ A2
    end

    subgraph B[eos_designs]
        direction LR
        B1{Input Processing<br/>Design Modules} --> B2[Intermediate<br/>Structured Config]
        B2 --> B4{Merge<br/>Overrides}
    end
```

The `eos_designs` role processes your design inputs in multiple stages:

1. **Input Processing & Design Modules**: Generate structured configuration based on your fabric design (topology, VLANs, VRFs, etc.)
2. **Intermediate Structured Config**: The auto-generated configuration following the EOS Config Data Model

> 3. **Merge Overrides**: Apply any device-specific `structured_config` (and/or `custom_structured_configuration_*` variables) on top of the generated configuration

Reviewer: Is it always `device-specific`?

Author: in the module when it is applied yes given that the inputs prsented to eos_designs_facts are device specific.

```yaml title="Input YAML"
# group_vars/FABRIC.yml
dns_settings:
  servers:
    - ip_address: 10.10.128.10
    - ip_address: 10.10.128.11

# host_vars/leaf1.yml
# Using custom_structured_configuration prefix
custom_structured_configuration_ip_name_servers:
  - ip_address: 10.20.20.10
    vrf: MGMT
  - ip_address: 10.10.128.10
    vrf: MGMT
    priority: 2
```

```mermaid
graph LR
    A1 --> B1
    A2 --> B4
    B4 --> C["ip_name_servers:<br/>&nbsp;&nbsp;- ip_address: 10.10.128.10<br/>&nbsp;&nbsp;&nbsp;&nbsp;vrf: MGMT<br/>&nbsp;&nbsp;&nbsp;&nbsp;priority: 2<br/>&nbsp;&nbsp;- ip_address: 10.10.128.11<br/>&nbsp;&nbsp;&nbsp;&nbsp;vrf: MGMT<br/>&nbsp;&nbsp;- ip_address: 10.20.20.10<br/>&nbsp;&nbsp;&nbsp;&nbsp;vrf: MGMT"]
    C --> D{eos_cli_config_gen}
    D --> E["ip name-server vrf MGMT 10.10.128.10 priority 2<br/>ip name-server vrf MGMT 10.10.128.11<br/>ip name-server vrf MGMT 10.20.20.10"]

    subgraph A[AVD Design Inputs]
        direction TB
        A1["dns_settings:<br/>&nbsp;&nbsp;servers:<br/>&nbsp;&nbsp;&nbsp;&nbsp;- 10.10.128.10<br/>&nbsp;&nbsp;&nbsp;&nbsp;- 10.10.128.11"]

>         A2["leaf1:<br/>&nbsp;&nbsp;custom_structured_configuration_ip_name_servers:<br/>&nbsp;&nbsp;&nbsp;&nbsp;- ip_address: 10.20.20.10<br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;vrf: MGMT<br/>&nbsp;&nbsp;&nbsp;&nbsp;- ip_address: 10.10.128.10<br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;vrf: MGMT<br/>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;priority: 2"]

Reviewer: this looks on the graph as if `custom_structured_config..` is under `leaf1` parent key and is not a root key

Author: indeed .. need to update the diagram

<!--
  ~ Copyright (c) 2026 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
-->

# Ansible Usage Patterns

This guide demonstrates the different ways to use Arista AVD Ansible roles, starting from basic usage and progressively introducing more advanced patterns. Each section builds upon the previous one, showing how AVD's flexibility allows you to customize configurations to meet your specific needs.

- [eos_cli_config_gen](#eos_cli_config_gen)
- [eos_designs with eos_cli_config_gen (recommended)](#eos_designs-with-eos_cli_config_gen)
  - [AVD Design Data Model with eos_designs](#avd-design-data-model-with-eos_designs)
  - [structured_config with eos_designs](#structured_config-with-eos_designs)
  - [custom_structured_configuration Prefix](#custom_structured_configuration-prefix)

## eos_cli_config_gen

The [eos_cli_config_gen](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md) role is the EOS config generation layer of AVD. It converts structured configuration data (in YAML format) into EOS configuration syntax.

> !!! warning "Warning"

Reviewer 2: this repeated "Warning" looks a bit awkward here

Author: while this already renders ok, using the following snippet will render the same (capitalize type of admonition) so lets simplify

```suggestion
!!! warning
```

### AVD Design Data Model with eos_designs

#### Workflow

```mermaid
graph LR
    A[AVD Design Data Model] --> B{eos_designs}
    B --> C[Structured Configuration<br/>EOS Config Data Model]
    C --> D{eos_cli_config_gen}
    D --> E[EOS CLI Configuration]
```

The [eos_designs](../../ansible_collections/arista/avd/roles/eos_designs/README.md) role implements Arista's best practices and design patterns. You provide high-level design parameters with the [AVD Design](../../ansible_collections/arista/avd/roles/eos_designs/docs/data-models.md) data model, like fabric topology, VLANs and VRFs, and the role will:

1. Generates the complete structured configuration following the [EOS Config](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/data-models.md) data model.

> 2. Passes this to [eos_cli_config_gen](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md) for EOS configuration generation.

Reviewer 2: ```suggestion
The [eos_designs](../../ansible_collections/arista/avd/roles/eos_designs/README.md) role implements Arista's best practices and design patterns. You provide high-level design parameters using the [AVD Design](../../ansible_collections/arista/avd/roles/eos_designs/docs/data-models.md) data model, like fabric topology, VLANs and VRFs, and the role will:

1. Generate the complete structured configuration following the [EOS Config](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/data-models.md) data model.
2. Pass this to [eos_cli_config_gen](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md) for EOS configuration generation.
```

### Key Takeaways

1. **Start with eos_designs**: Most users should begin with [eos_designs](../../ansible_collections/arista/avd/roles/eos_designs/README.md) for fabric deployments, not [eos_cli_config_gen](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md) directly.
2. **Override When Needed**: Use `structured_config` or `custom_structured_configuration_` for exceptions.
3. **Combine Approaches**: The last two approaches can be used together - both `structured_config` and `custom_structured_configuration_*` variables are merged during the same step.
4. **Understand Precedence**: Custom configuration is merged *after* design generation, allowing overrides.
5. **Choose Your Style**: Pick the override method that fits your workflow - nested `structured_config` or prefixed variables.

### Next Steps

Review the following resources to learn more:

- [Custom Structured Configuration How-To](../../ansible_collections/arista/avd/roles/eos_designs/docs/how-to/custom-structured-configuration.md) for advanced patterns.
- [AVD Design](../../ansible_collections/arista/avd/roles/eos_designs/docs/data-models.md) data model for available design options.

> - [EOS Config](../../ansible_collections/arista/avd/roles/eos_designs/README.md) data model for available EOS configuration options.

Reviewer: ```suggestion
- [EOS Config](../../ansible_collections/arista/avd/roles/eos_cli_config_gen/docs/data-models.md) data model for available EOS configuration options.
```

Reviewer: @gmuloc
